File: 30-days-of-code/week 1/D6.GroupAnagrams.swift
/*
Given an array of strings, group anagrams together.

Example:

Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
Output:
[
  ["ate","eat","tea"],
  ["nat","tan"],
  ["bat"]
]
Note:

All inputs will be in lowercase.
The order of your output does not matter.
*/
class Solution {
    func groupAnagrams(_ strs: [String]) -> [[String]] {
        var emptyDictionary:[String:[String]] = [:]
        
        for str in strs{
            var ordered = String(str.sorted())
            print(ordered)
            if(emptyDictionary[ordered] != nil){
                emptyDictionary[ordered]?.append(str) 
            }else{
                emptyDictionary[ordered] = [str]
            }
        }
        
        return Array(emptyDictionary.values)
    }
}

File: 30-days-of-code/week 1/D6.GroupAnagrams.java
/*
Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

 

Example 1:

Input: strs = ["eat","tea","tan","ate","nat","bat"]
Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
Example 2:

Input: strs = [""]
Output: [[""]]
Example 3:

Input: strs = ["a"]
Output: [["a"]]
*/

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        
        for(String str : strs){
            char[] chars = str.toCharArray();
            Arrays.sort(chars);
            String sorted = new String(chars);

            if(map.containsKey(sorted)){
                List temp = map.get(sorted);
                temp.add(str);
            }else{
                ArrayList<String> tempArray = new ArrayList<String>();
                tempArray.add(str);
                map.put(sorted, tempArray);
            }
        }
        
        return new ArrayList<>(map.values());
    }
}